kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaProtocolHandler.java

```diff
@@ -464,4 +459,66 @@ private void loadTxnLogTopics(TransactionCoordinator txnCoordinator) throws Exce
             log.info("Current broker: {} does not own any of the txn log topic partitions", currentBroker);
         }
     }
+
+    private static CompletableFuture<InetSocketAddress> getFailedAddressFuture(final Throwable throwable) {
+        final CompletableFuture<InetSocketAddress> future = new CompletableFuture<>();
+        future.completeExceptionally(throwable);
+        return future;
+    }
+
+    private static CompletableFuture<InetSocketAddress> getAddressFutureFromBrokerUrl(final String brokerUrl) {
+        final CompletableFuture<InetSocketAddress> future = new CompletableFuture<>();
+        try {
+            final URI uri = new URI(brokerUrl);
+            future.complete(InetSocketAddress.createUnresolved(uri.getHost(), uri.getPort()));
+        } catch (URISyntaxException e) {
+            future.completeExceptionally(e);
+        }
+        return future;
+    }
+
+    public static CompletableFuture<InetSocketAddress> getBroker(final PulsarService pulsarService,
+                                                                 final TopicName topicName) {
+        final NamespaceService namespaceService = pulsarService.getNamespaceService();
+        if (namespaceService == null) {
+            return getFailedAddressFuture(new IllegalStateException("NamespaceService is not available"));
+        }
+
+        final LookupOptions options = LookupOptions.builder()
+                .authoritative(false)
+                .advertisedListenerName(null)
+                .loadTopicsInBundle(true)
+                .build();
+        final CompletableFuture<InetSocketAddress> future =
+                namespaceService.getBrokerServiceUrlAsync(topicName, options).thenCompose(optLookupResult -> {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Lookup result {}", topicName.toString(), optLookupResult);
+            }
+            if (!optLookupResult.isPresent()) {
+                return getFailedAddressFuture(ClientCnx.getPulsarClientException(
+                        ServerError.ServiceNotReady,
+                        "No broker was available to own " + topicName));
+            }
+
+            final LookupResult lookupResult = optLookupResult.get();
+            if (lookupResult.isRedirect()) {
```

Comment:
@codelipenghui In this case, it seems there's no good way to handle it. I've tried two ways before.
1. Complete it with the broker url of lookup data. However, then Kafka client will send a subsequent PRODUCE request to a non-owner broker.
2. Complete it exceptionally just like the case `optLookupResult` is not present. Then Kafka client still failed.
They can both be reproduced in KoP's `DistributedClusterTest`, which run two mocked broker.

Reply:
The reason might be that if we completed the future immediately, the actual owner broker's ownership cache could not be updated without an extra request from client side.

Reply:
the candidate broker is already in `lookupResult` here, can we create a pulsar client with the candidate broker url to send the extra request with `authoritative = true`?
another way we could return the candidate broker url as lookup result to Kafka client, Kafka client will send next request to the candidate broker, we can update the ownership cache there in the next request handler.

Reply:
> can we create a pulsar client with the candidate broker url to send the extra request with authoritative = true

Sorry I don't understand well. Here I used `PulsarAdmin` to do it. Can you explain the difference between `PulsarClient` and `PulsarAdmin` for topic lookup? And what does `authoritative = true` means?

> another way we could return the candidate broker url as lookup result to Kafka client

This way doesn't work. If we return the candicate broker url to Kafka client, the Kafka client won't send another round of METADATA request. Instead, it will send PRODUCE request to that broker directly, then `NOT_LEADER_FOR_PARTITION` happens.

Reply:
1. `admin.lookups().lookupTopicAsync()` currently does not support listener name, so we can not set listener name here to do topic lookup. Listener name is supported by HttpLookupService by https://github.com/apache/pulsar/pull/10319 but admin api miss to add the parameter.
2. Http lookup will call `TopicLookupBase.internalLookupTopicAsync`，which set `loadTopicsInBundle=false`, but we set `.loadTopicsInBundle(true)` before, I'm not sure if this will cause any difference in future operation. 
3. `authoritative = true` means this is a redirect request, the broker which handle the request is the candidate broker to own the topic. otherwise loadmanager will select a candidate broker again.

Reply:
Thanks for you detail explanation very much! I'll apply your suggestions soon.